### PR TITLE
Packaging pipeline

### LINKED ISSUE #2454: Release and packaging pipeline
We have a couple of "issues" with our current setup:

- #2452 had to introduce a hack to install a Sphinx plugin. This is an consequence of running docs build as part of `conda build`.
- `meta.yaml` defines a long list of "build" requirements that are actually requirements for the docs build.
- Wheel and conda package builds a 2 separate workflows. Should consider unifying this such that uploads happen only if all pass.
- Docs build could be done on packaged scipp, as a separate job.
- No one uses `scipp/label/dev` conda packages. This is a waste of resources.
- Docs get updated when a new "dev" package is made, i.e., does not reflect state the users sees when installing scipp

Incomplete suggestion:

- Do not make `scipp/label/dev` packages any more.
- Remove docs build from `meta.yaml`.
- Merge wheel and conda build workflow into a single workflow.
- Add a job to build the docs based on a package made in one of the previous jobs (conda or pip). Could consider publishing conda packages to a `staging`/`rc` label for this purpose, and relabel to `main` only if docs build passes.

### PR DESCRIPTION
**Pull request by nvaytet**
This is a first attempt at solving some of the issues outlined in #2454.

- We do not make dev packages anymore. The weekly jobs have gone, we only run conda-build on releases.

- Moved the docs building to a separate job:
  - The conda package is first built, and uploaded to anaconda under a `staging` label.
  - The docs building dependencies have moved out of the `conda/meta.yaml` file and into the `developer-extra.yml` file, which is combined with the `conda/meta.yaml` to create a environment for building the docs.
  - The `staging` package is then installed in a conda environment to build the docs in a second job

- If docs build is successful, re-label the conda package from `staging` to `main` using anaconda client

Fixes #2454

### CODE REVIEW

**Review by jl-wynen — APPROVED** [review]

**Comment by SimonHeybrock** [comment]
Before having looked at the code:

>       * The conda package is first built, and uploaded to anaconda under a `staging` label.
>       * The docs building dependencies have moved out of the `conda/meta.yaml` file and into the `developer-extra.yml` file, which is combined with the `conda/meta.yaml` to create a environment for building the docs.

Have you considered using `pip` to install scipp for the docs build? Then you do not need to mess around with conda and files and reduce the reliance and custom and brittle scripts?

File: .github/workflows/pr_and_main.yml [hunk 1]
```diff
@@ -85,9 +85,7 @@ jobs:
       - run: ctest --preset test
       - run: python -m pytest -v tests
 
-      - run: |
-          pip install sphinx-autodoc-typehints>=1.17.0
-          cmake --build --preset docs
+      - run: cmake --build --preset docs
```
Comment (SimonHeybrock) [inline .github/workflows/pr_and_main.yml:88]: So all systems install the docs dependencies (as before), but only one build uses it (this was also the case before, just asking to clarify)?

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+  deploy:
+    needs: docs
+    name: Deploy docs and re-label conda packages to main
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2  # Need to checkout repo so github-pages-deploy-action works
+      - uses: actions/download-artifact@v2
+
+      # TODO: can we have multiple target folders so that we don't have the action twice?
```
Comment (SimonHeybrock) [inline .github/workflows/release.yml:63]: Use `cp` and just deploy the root folder?

**Review by SimonHeybrock — COMMENTED** [review]
What about the wheel builds for PyPI? These have previously failed independently of the conda builds, so we should also try to integrate them here.

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+          repository-name: scipp/scipp.github.io
+          branch: master
+          folder: html
+          target-folder: ${{ format('release/{0}', github.ref_name) }}
+          single-commit: true
+          clean-exclude: release
+          ssh-key: ${{ secrets.GH_PAGES_DEPLOY_KEY }}
+
+      # Change conda packages labels from staging to main
+      - uses: conda-incubator/setup-miniconda@v2
```
Comment (SimonHeybrock) [inline .github/workflows/release.yml:87]: Probably cleaner to make this a separate job?

**Comment by nvaytet** [comment]
Looks good to me, I cannot click the approve button because I originally opened the PR

**Pull request merged**